src/middleware/insights.ts

```diff
@@ -0,0 +1,121 @@
+import { Middleware } from '.';
+import { InsightsClient, InstantSearch } from '../types';
+import { getInsightsAnonymousUserToken } from '../helpers';
+
+export type InsightsProps = {
+  insightsClient: InsightsClient;
+};
+
+export type CreateInsightsMiddleware = (props: InsightsProps) => Middleware;
+
+export const createInsightsMiddleware: CreateInsightsMiddleware = props => {
+  const { insightsClient } = props;
+  if (!insightsClient) {
+    throw new Error(
+      'passing insightsClient to instantsearch is required for insightsMiddleware'
+    );
+  }
+  return ({ instantSearchInstance }) => {
+    const setUserToken = (userToken?: string) => {
+      // At the time this middleware is subscribed, `mainIndex.init()` is already called.
+      // It means `mainIndex.getHelper()` exists.
+      if (userToken) {
+        instantSearchInstance.mainIndex
+          .getHelper()!
+          .setQueryParameter('userToken', userToken);
+      }
+    };
+
+    return {
+      onStateChange() {},
+      subscribe() {
+        updateQueryParameter(instantSearchInstance);
+        ensureInitialization({
+          instantSearchInstance,
+          insightsClient,
+          callback: () => {
+            setUserTokenFromQueueOrCookie(insightsClient, setUserToken);
+            watchUserToken(insightsClient, setUserToken);
+          },
+        });
```

Comment:
Why so many indirections? It makes the code harder to understand and to delete in the end. We better inline most of it—only `getAppIdAndApiKey` seems useful to extract.

Reply:
@francoischalifour 
Okay, here, no offence at all, but genuinely curious on your perspective:

* Why do you think they're indirect?
* Maybe I have abstracted them in a poor way?
* Or, names aren't good enough to help you understand?

I splited the code into relevant chunks and named them after what they do. 

And I just tried inlined them after you comment, and I'm not sure if they're better.

* Do you think this inlined version is better than the abstracted one?
* Or, in your mind, did you have a better version than this screenshot?

![image](https://user-images.githubusercontent.com/499898/88375746-6b7a7080-cd9c-11ea-81f9-f0720e78b56a.png)

Reply:
In the current PR, you created functions just to name what a piece of code was doing. This is not really the point of functions—they're aimed at reducing complexity when the code they're hosting is supposed to be used at several places.

The code on the screenshot is way more straightforward. Later on, we'll feel confident removing a piece of code because it's inlined. Creating functions brings a mental overhead that creates friction when times comes to refactor/delete code (e.g., "this function had a name, we probably knew what we meant with this, let's not remove it").

From the outside, we (the team, the contributors, and yourself in a few days) understand the inlined code way better because it reads from top to bottom easily—with the functions, we jump from one statement to another, loosing track of what gets executed and when it gets executed.

I'd recommend reading "[Goodbye, Clean Code](https://overreacted.io/goodbye-clean-code/)".

Reply:
Thanks for the explanation and the link. I'll give this change a try!